Upgrade to Littlefs rev2.4.1

poojamg opened the pull request.

Commit:
lfs_bd_cmp() compares more bytes at one time

It's very slowly to compare one byte at one time. Here are the
performance I get from 128M spinand with NFTL by sequential writing.

| file size | buffer size  | write speed  |
| 10 MB     | 0   B        | 3206.01 KB/s |
| 10 MB     | 1   B        | 2434.04 KB/s |
| 10 MB     | 2   B        | 2685.78 KB/s |
| 10 MB     | 4   B        | 2857.94 KB/s |
| 10 MB     | 8   B        | 3060.68 KB/s |
| 10 MB     | 16  B        | 3155.30 KB/s |
| 10 MB     | 64  B        | 3193.68 KB/s |
| 10 MB     | 128 B        | 3230.62 KB/s |
| 10 MB     | 256 B        | 3153.03 KB/s |

| 70 MB     | 0   B        | 2258.87 KB/s |
| 70 MB     | 1   B        | 1827.83 KB/s |
| 70 MB     | 2   B        | 1962.29 KB/s |
| 70 MB     | 4   B        | 2074.01 KB/s |
| 70 MB     | 8   B        | 2147.03 KB/s |
| 70 MB     | 64  B        | 2179.92 KB/s |
| 70 MB     | 256 B        | 2179.96 KB/s |

The 0 Byte size means no validation and the 1 Byte size is how
littlefs do before. Based on the above table and to save memory,
comparing 8 bytes at one time is more wonderful.

Signed-off-by: WeiXiong Liao <liaoweixiong@allwinnertech.com>

rojer committed:
Fix -Wmissing-field-initializers

rojer committed:
Remove -Wno-missing-field-initializers

geky committed:
Merge pull request #405 from rojer/mfe

Fix -Wmissing-field-initializers

jrast committed:
Added littlefs-python to the related projects section

As introduced in #297, I created a python wrapper for littlefs. The wrapper supports two API's: A C-like API which is the same as in C and a more pythonic API which is easier to use if you are more the python guy. The wrapper is built with littlefs 2.2.1 at the moment.

Commit:
Assert that file/dir struct is not reused in lfs_file_opencfg/lfs_dir_open

Commit:
Use LFS_O_RDWR instead of magic number in lfs_file_* asserts

sgupta20 committed:
Fix assert

renesas-billgesner committed:
add thread safe wrappers

renesas-billgesner committed:
Squash of thread-safe PR cleanup

- expand functions
- add comment
- rename functions
- fix locking issue in format and mount
- use global include
- fix ac6 linker issue
- use the global config file
- address review comments
- minor cleanup
- minor cleanup
- review comments

maximevince committed:
Add LFS_READYONLY define, to allow smaller builds providing read-only mode

geky committed:
Fixed incorrect modulus in lfs_alloc_reset

Modulus of the offset by block_size was clearly a typo, and should be
block_count. Interesting to note that later moduluses during alloc
calculations prevents this from breaking anything, but as gtaska notes it
could skew the wear-leveling distribution.

Found by guiserle and gtaska

maximevince committed:
Squash of LFS_READONLY cleanup

- undef unavailable function declarations altogether
- even less code, assert on write attempts
- remove LFS_O_WRONLY and other flags when compiling with LFS_READONLY
- do not annotate #endif, as requested
- move ifdef before comments blocks, rework dangling opening bracket
- ifdef file flags that are not needed in read-only mode
- slight refactor
- ifdef LFS_F_ERRED out as well

Commit:
Add lfs_mlist_append/remove helper

Commit:
Deprecate LFS_F_OPENED and use lfs_mlist_isused instead

Instead of additional flag, we can just go through the mlist.

geky committed:
Removed unnecessary randomization of offsets in lfs_alloc_reset

On first read, randomizing the allocators offset may seem appropriate
for lfs_alloc_reset. However, it ends up using the filesystem-fed
pseudorandom seed in situations it wasn't designed for.

As noted by gtaska, the combination of using xors for feeding the seed
and multiple traverses of the same CRCs can cause the seed to flip to
zeros with concerning frequency.

Removed the randomization from lfs_alloc_reset, leaving it in only
lfs_mount.

Found by gtaska

geky committed:
Switched to CRC as seed collection function instead of xor

As noted by gtaska, we are sitting on a better hash-combining function
than xor: CRC. Previous issues with xor were solvable, but relying on
xor for this isn't really worth the risk when we already have a CRC
function readily available.

To quote a study found by gtaska:

https://michiel.buddingh.eu/distribution-of-hash-values

> CRC32 seems to score really well, but its graph is skewed by the results
> of Dataset 5 (binary numbers), which may or may not be too synthetic to
> be considered a fair benchmark. But even if you substract the results
> from that test, it does not fare significantly worse than other,
> cryptographic hash functions.

renesas-billgesner committed:
make raw functions static. formatting tweaks

geky committed:
Added read-only build+size reporting to CI

geky committed:
Fixed allocation-eviction issue when erase state is multiple of block_cycles+1

This rather interesting corner-case arises in lfs_dir_alloc anytime the
uninitialized revision count happens to be a multiple of block_cycles+1.

For example, the source of the bug found by tim-nordell-nimbelink:

rev = 2742492087
block_cycles = 100

2742492087 % (100+1) = 0

The reason for this weird block_cycles+1 case is due to a fix for a
previous bug in fe957de. To avoid aliasing, which would cause metadata
pairs to wear unevenly, block_cycles incremented to the next odd number.

Normally, littlefs tweaks the revision count of blocks during
lfs_dir_alloc in order to make sure evictions can't happen on the first
compact. Otherwise, higher-level logic such as lfs_format would break.

However, this wasn't updated with the aliasing fix in fe957de, so
lfs_dir_alloc was only rounding the revision count to the nearest even
number.

The current fix is to change the logic in lfs_dir_alloc to explicitly
check for the eviction condition and increment if eviction would occur.

Found by tim-nordell-nimbelink

geky committed:
Fixed typo in scripts/readtree.py

Not sure how this went unnoticed, I guess this is the first bug that
needed in-depth inspection after the a last-minute argument cleanup
in the debug scripts.

geky committed:
Fixed single unchecked bit during commit verification

This bug was exposed by the bad-block tests due to changes to block
allocation, but could have been hit before these changes.

In flash, when blocks fail, they don't fail in a predictable manner. To
account for this, the bad-block tests check a number of failure
behaviors. The interesting one here is "LFS_TESTBD_BADBLOCK_ERASENOOP",
in which bad blocks can not be erased or programmed, and are stuck with
the data written at the time the blocks go bad.

This is actually a pretty realistic failure behavior, since flash needs a
large voltage to force the electrons of the floating gates. Though
realistically, such a failure would like corrupt the data a bit, not leave the
underlying data perfectly intact.

LFS_TESTBD_BADBLOCK_ERASENOOP is rather interesting to test for because it
means bad blocks can end up with perfectly valid CRCs after a failed write,
confusing littlefs.

---

In this case, we had the perfect series of operations such that a test
was repeatedly writing the same sequence of metadata commits to the same
block, which eventually goes bad, leaving the block stuck with metadata
that occurs later in the sequence.

What this means is that after the first commit, the metadata block
contained both the first and second commits, even though the loop in the
test hadn't reached that point yet.

expected       actual
.----------.  .----------.
| commit 1 |  | commit 1 |
| crc 1    |  | crc 1    |
|          |  | commit 2 <-- (from previous iteration)
|          |  | crc 2    |
'----------'  '----------'

To protect against this, littlefs normally compares the written CRC
against the expected CRC, but because this was the exact same data that
it was going to write, this CRCs end up the same.

Ah! But doesn't littlefs also encode the state of the next page to keep
track of if the next page has been erased or not? Wouldn't that change
between iterations?

It does! In a single bit in the CRC-tag. But thanks to some incorrect
logic attempting to avoid an extra condition in the loop for writing out
padding commits, the CRC that littlefs checked against was the CRC
immediately before we include the "is-next-page-erased" bit.

Changing the verification check to use the same CRC as what is used to
verify commits on fetch solves this problem.

geky committed:
Tweaked thread-safe implementation

- Stayed on non-system include for lfs_util.h for now
- Named internal functions "lfs_functionraw"
- Merged lfs_fs_traverseraw
- Added LFS_LOCK/UNLOCK macros
- Changed LFS_THREADSAFE from 1/0 to defined/undefined to
  match LFS_READONLY

geky committed:
Moved LFS_TRACE calls to API wrapper functions

This removes quite a bit of extra code needed to entertwine the
LFS_TRACE calls into the original funcions.

Also changed temporary return type to match API declaration where
necessary.

geky committed:
Added thread-safe build+size reporting to CI

geky committed:
Changed lfs_dir_alloc to maximize block cycles for new metadata pairs

Previously we only bumped the revision count if an eviction would occur
immediately (and possibly corrupt littlefs). This works, but does risk
an unoptimal superblock size if an almost-exhausted superblock was
allocated during lfs_format.

As pointed out by tim-nordell-nimbelink, we can align the revision count
to maximize the number of block cycles without breaking the existing
requirements of increasing revision counts.

As an added benefit, littlefs's wear-leveling should behave more
consistently after this change.

geky committed:
Merge pull request #412 from jrast/patch-3

Added littlefs-python to the related projects section

geky committed:
Merge pull request #486 from littlefs-project/fix-assert

Fix assert

geky committed:
Merge pull request #487 from littlefs-project/fix-alloc-reset-modulus

Fix several wear-leveling issues found in lfs_alloc_reset

geky committed:
Merge pull request #395 from gmpy/improve-write-performance

lfs_bd_cmp() compares more bytes at one time

geky committed:
Merge pull request #443 from NoahGorny/add-already-opened-assert

Assert that the file isnt open in lfs_file_opencfg

geky committed:
Merge pull request #480 from maximevince/master

Add LFS_READONLY define, to allow smaller builds providing read-only mode

geky committed:
Merge pull request #470 from renesas/SWFLEX-1517-littlefs-thread-safe-option

Add thread safe wrappers

geky committed:
Moved lfs_mlist_isopen checks into the API wrappers

This indirectly solves an issue with lfs_file_rawclose asserting
when lfs_file_opencfg errors since appending to the mlist occurs
after open. It also may speed up some of the internal operations such as
the lfs_file_write used to resolve unflushed data.

The idea behind adopting mlist over flags is that realistically it's
unlikely for the user to open a significant number of files (enough for
big O to kick in). That being said, moving the mlist asserts into the
API wrappers does protect some of the internal operations from scaling
based on the number of open files.

geky committed:
Merge pull request #490 from littlefs-project/fix-alloc-eviction

Fix allocation-eviction issue when erase state is multiple of block_cycles+1

geky committed:
Bumped minor version to v2.3

geky committed:
Renamed internal lfs_*raw -> lfs_raw* functions

- Prefixing with raw is slightly more readable, follows
  common-prefix rule
- Matches existing raw prefixes in testbd

geky committed:
Merge pull request #495 from littlefs-project/devel

Minor release: v2.3

geky committed:
Moved lfs_mdir_isopen behind LFS_NO_ASSERT

lfs_mdir_isopen goes unused if asserts are disabled, and this caused an
"unused function" warning on Clang (curiously not on GCC since the
function was static inline, commonly used for header-only functions).

Also removed "inline" from the lfs_mdir_* functions as these involve
linked-list operations and really shouldn't be inlined. And since they
are static, inlining should occur automatically if there is a benefit.

Found by dpgeorge

geky committed:
Added scripts/code_size.py, for more in-depth code-size reporting

Inspired by Linux's Bloat-O-Meter, code_size.py wraps nm to provide
function-level code size, and supports detailed comparison between
different builds.

One difference is that code_size.py invokes littlefs's build system
similarly to test.py, creating a duplicate build in the "sizes"
directory. This makes it easy to monitor a cross-compiled build size
while simultaneously testing on the host machine.

geky committed:
Fixed incorrect documentation in test.py

The argparse documented an outdated format, and was off by 1.

Found by sender6

mon committed:
Add metadata_max and inline_file_max to config

We have seen poor read performance on NAND flashes with 128kB blocks.
The root cause is inline files having to traverse many sets of metadata
pairs inside the current block before being fully reconstructed. Simply
disabling inline files is not enough, as the metadata will still fill up
the block and eventually need to be compacted.

By allowing configuration of how much size metadata takes up, along with
limiting (or disabling) inline file size, we achieve read performance
improvements on an order of magnitude.

mon committed:
Remove inline_files_max and lfs_t entry for metadata_max

geky committed:
Added GitHub workflows to run tests

Mostly taken from .travis.yml, biggest changes were around how to get
the status updates to work.

We can't use a token on PRs the same way we could in Travis, so instead
we use a second workflow that checks every pull request for "status"
artifacts, and create the actual statuses in the "workflow_run" event,
where we have full access to repo secrets.

geky committed:
Moved sanity check in lfs_format after compaction

After a bit of tweaking in 9dde5c7 to write out all superblocks
during lfs_format, additional writes were added after the sanity
checking normally done at the end.

This turned out to be a problem when porting littlefs, as it makes it
easy for addressing issues to not get caught during lfs_format.

Found by marekr, tristanclare94, and mjs513

embeddedt committed:
Skip flushing file if lfs_file_rawseek() doesn't change position

geky committed:
Added coverage.py, and optional coverage info to test.py

Now coverage information can be collected if you provide the --coverage
to test.py. Internally this uses GCC's gcov instrumentation along with a
new script, coverage.py, to parse *.gcov files.

The main use for this is finding coverage info during CI runs. There's a
risk that the instrumentation may make it more difficult to debug, so I
decided to not make coverage collection enabled by default.

geky committed:
Switched to lcov for coverage collection, greatly simplified coverage.py

Since we already have fairly complicated scriptts, I figured it wouldn't
be too hard to use the gcov tools and directly parse their output. Boy
was I wrong.

The gcov intermediary format is a bit of a mess. In version 5.4, a
text-based intermediary format is written to a single .gcov file per
executable. This changed sometime before version 7.5, when it started
writing separate .gcov files per .o files. And in version 9 this
intermediary format has been entirely replaced with an incompatible json
format!

Ironically, this means the internal-only .gcda/.gcno binary format has
actually been more stable than the intermediary format.

Also there's no way to avoid temporary .gcov files generated in the
project root, which risks messing with how test.py runs parallel tests.
Fortunately this looks like it will be fixed in gcov version 9.

---

Ended up switching to lcov, which was the right way to go. lcov handles
all of the gcov parsing, provides an easily parsable output, and even
provides a set of higher-level commands to manage coverage collection
from different runs.

Since this is all provided by lcov, was able to simplify coverage.py
quite a bit. Now it just parses the .info files output by lcov.

geky committed:
Added BUILDDIR, a bit of script reworking

Now littlefs's Makefile can work with a custom build directory
for compilation output. Just set the BUILDDIR variable and the Makefile
will take care of the rest.

make BUILDDIR=build size

This makes it very easy to compare builds with different compile-time
configurations or different cross-compilers.

This meant most of code.py's build isolation is no longer needed,
so revisted the scripts and cleaned/tweaked a number of things.

Also bought code.py in line with coverage.py, fixing some of the
inconsistencies that were created while developing these scripts.

One change to note was removing the inline measuring logic, I realized
this feature is unnecessary thanks to GCC's -fkeep-static-functions and
-fno-inline flags.

geky committed:
Fixed a recompilation issue in CI, tweaked coverage.py a bit more

This was lost in the Travis -> GitHub transition, in serializing some of
the jobs, I missed that we need to clean between tests with different
geometry configurations. Otherwise we end up running outdated binaries,
which explains some of the weird test behavior we were seeing.

Also tweaked a few script things:
- Better subprocess error reporting (dump stderr on failure)
- Fixed a BUILDDIR rule issue in test.py
- Changed test-not-run status to None instead of undefined

mon committed:
lfs_fs_preporphans ret int for graceful LFS_ASSERT

geky committed:
Brought over the release workflow

This is pretty much a cleaned up version of the release script that ran
on Travis.

This biggest change is that now the release script also collecs the
build results into a table as part of the change notes, which is a nice
addition.

geky committed:
Reduced build sources to just the core littlefs

Currently this is just lfs.c and lfs_util.c. Previously this included
the block devices, but this meant all of the scripts needed to
explicitly deselect the block devices to avoid reporting build
size/coverage info on them.

Note that test.py still explicitly adds the block devices for compiling
tests, which is their main purpose. Humorously this means the block
devices will probably be compiled into most builds in this repo anyways.

geky committed:
Added post-release script, cleaned up workflows

This helps an outstanding maintainer annoyance: updating dependencies to
bring in new versions on each littlefs release.

But instead of adding a bunch of scripts to the tail end of the release
workflow, the post-release script just triggers a single
"repository_dispatch" event in the newly created littlefs.post-release
repo. From there any number of post-release workflows can be run.

This indirection should let the post-release scripts move much quicker
than littlefs itself, which helps offset how fragile these sort of scripts
are.

---

Also finished cleaning up the workflows now that they are mostly
working.

embeddedt committed:
Use lfs_file_rawsize to calculate LFS_SEEK_END position

mon committed:
Add test to ensure asserts can return

geky committed:
Removed .travis.yml

Now that it's been replaced by GitHub workflows (in .github/workflows)

geky committed:
Fixed lfs_file_truncate issue where internal state may not be flushed

This was caused by the new lfs_file_rawseek optimization that can skip
flushing when calculated file->pos is unchanged combined with an
implicit expectation in lfs_file_truncate that lfs_file_rawseek
unconditionally sets file->pos.

Because of this assumption, lfs_file_truncate could leave file->pos in
an outdated state while changing the internal file metadata. Humorously,
this was always gauranteed to trigger the skip in lfs_file_rawseek when
we try to restore the file->pos, leaving the file->cache used to do the
CTZ skip-list lookup in a potentially bad state.

The easiest fix is to just update file->pos correctly. Note we don't
want to explicitly flush since we can leverage the same noop
optimization if we truncate to the file position. Which I've added a
test for.

geky committed:
Removed unnecessary truncate condition thanks to new seek optimization

geky committed:
Merge pull request #502 from mon/feature/meta_limits

Add metadata_max config to help performance on devices with large blocks

geky committed:
Merge pull request #508 from littlefs-project/fix-sanity-check

Moved sanity check in lfs_format after compaction

geky committed:
Merge pull request #511 from embeddedt/fix_lseek

Skip flushing file if lfs_file_rawseek() doesn't change position

geky committed:
Merge pull request #514 from mon/feature/assert_early_return

lfs_fs_preporphans: return int to alllow graceful LFS_ASSERT

geky committed:
Merge branch 'devel' into ci-revamp

Needed to bring in new "error-asserts" configuration

geky committed:
Enabled overriding of LFS_ASSERT/TRACE/DEBUG/etc

This is useful for testing the new erroring assert behavior in CI.
Asserts do not error by default, so this macro needs to be overriden.

It is possible to test this behavior using the existing option of
overriding lfs_util.h with a custom file, by using a small sed
one-line script. But this is much simpler.

This does raise the question if more of the configuration options in
lfs_util.h should be opened up for function-like macro overrides.

geky committed:
Merge pull request #516 from littlefs-project/ci-revamp

Adopt GitHub Actions, bring in a number of script/Makefile improvements

geky committed:
Bumped minor version to v2.4

geky committed:
Merge pull request #519 from littlefs-project/devel

Minor release: v2.4

yamt committed:
scripts/test.py: Fix infinite busy loops on macOS

I confirmed that the same number of tests are run
with "make test" on:

    * Ubuntu with and without this change
    * macOS with this change

>   ====== results ======
>   tests passed 817/817 (100.00%)
>   tests failed 0/817 (0.00%)

tniessen committed:
Fix link to test block device

geky committed:
Merge pull request #529 from yamt/macos-make-test

scripts/test.py: Fix infinite busy loops on macOS

geky committed:
Merge pull request #565 from tniessen/fix-link-to-test-bd

Fix link to test block device

Commit:
Merge branch 'master' of https://github.com/littlefs-project/littlefs into HEAD

dleach02 commented on .github/workflows/post-release.yml:
```diff
@@ -0,0 +1,26 @@
+name: post-release
+on:
```
Not sure you intended to include these .github files

poojamg replied:
My bad have removed .github files ..

dleach02 commented:
You will need a PR in Zephyr updating the west.yml

poojamg commented:
> You will need a PR in Zephyr updating the west.yml
Thanks @dleach02, I have added it now
https://github.com/zephyrproject-rtos/zephyr/pull/40484

Commit:
Fix lfs_mlist_isopen unused-function error

Fix compiler error ‘lfs_mlist_isopen’ defined but not used'

Signed-off-by: Pooja Ganesh <poojamg@amperecomputing.com>

Commit:
Fix redefinition of LOG module

Fix redefinition of ‘__log_level’ already declared at
LOG_MODULE_REGISTER(settings_basic_test); bcoz of LOG module declare in
lfs_util.h

Signed-off-by: Pooja Ganesh <poojamg@amperecomputing.com>

tejlmand commented:
@poojamg this being a merge commit bringing Zephyr fork up-to-data with v2.4.1, then why does the Zephyr fork still contain:
https://github.com/zephyrproject-rtos/littlefs/blob/zephyr/.travis.yml

which were removed here:
https://github.com/zephyrproject-rtos/littlefs/commit/6592719d28d8f3af931680cc04f77cf855a0b3b8

Wrong decision when resolving conflicts ?

de-nordic commented:
> @poojamg this being a merge commit bringing Zephyr fork up-to-data with v2.4.1, then why does the Zephyr fork still contain: https://github.com/zephyrproject-rtos/littlefs/blob/zephyr/.travis.yml
> 
> which were removed here: [6592719](https://github.com/zephyrproject-rtos/littlefs/commit/6592719d28d8f3af931680cc04f77cf855a0b3b8)
> 
> Wrong decision when resolving conflicts ?

The file does not show on the diff of this PR so I do not know where it comes from.

poojamg commented:
> > @poojamg this being a merge commit bringing Zephyr fork up-to-data with v2.4.1, then why does the Zephyr fork still contain: https://github.com/zephyrproject-rtos/littlefs/blob/zephyr/.travis.yml
> > which were removed here: [6592719](https://github.com/zephyrproject-rtos/littlefs/commit/6592719d28d8f3af931680cc04f77cf855a0b3b8)
> > Wrong decision when resolving conflicts ?
> 
> The file does not show on the diff of this PR so I do not know where it comes from.

Hmm, I don't see which commit added back this file, I don't understand one commits delete and succeeding commit modifies the file (I don't see adding it back commit) 
![image](https://user-images.githubusercontent.com/2581319/148979048-b33def8e-5992-41b2-b3f4-01c7cc02063e.png)

tejlmand commented:
> The file does not show on the diff of this PR so I do not know where it comes from.

It does, try fetch this PR and diff it against `v2.4.1` for example:
```
$ git fetch zephyr pull/7/head
$ git diff v2.4.1 FETCH_HEAD -- .travis
diff --git a/.travis.yml b/.travis.yml
new file mode 100644
index 0000000..78d964a
--- /dev/null
+++ b/.travis.yml
@@ -0,0 +1,429 @@
+# environment variables
+env:
+  global:
+    - CFLAGS=-Werror
+    - MAKEFLAGS=-j
+
+# cache installation dirs
+cache:
+  pip: true
+  directories:
+    - $HOME/.cache/apt
+
+# common installation
....
```

I fixed it here: https://github.com/zephyrproject-rtos/littlefs/pull/9

tejlmand commented:
>  I don't understand one commits delete and succeeding commit modifies the file 

Exactly, so because one branch removes the file and the other modifies it you should have gotten a merge conflict.
And `git` should have asked during conflict resolving whether you want to keep the deleted or modified file.

This is why I wrote: `Wrong decision when resolving conflicts ?` earlier.
